File: python/src/contract_filter.py
"""
Sistema de filtragem e classificação de contratos
"""

import logging
import re
from typing import List, Dict, Any

logger = logging.getLogger(__name__)


class ContractFilter:
    """Filtra e classifica contratos com base no perfil do utilizador"""

    def __init__(self, profile: Dict[str, Any]):
        self.keywords = [kw.lower() for kw in profile.get('keywords', [])]
        self.exclude_terms = [term.lower() for term in profile.get('excludeTerms', [])]
        self.min_price = profile.get('minPrice', 0)
        self.max_price = profile.get('maxPrice', float('inf'))

    def _parse_price(self, price_str: str) -> float:
        """
        Converter preço "75.000,00 €" para 75000.00
        """
        try:
            # Remover símbolo €, espaços, e converter formato PT para float
            cleaned = price_str.replace('€', '').replace(' ', '').strip()
            # Substituir separador de milhares (.) e decimal (,)
            cleaned = cleaned.replace('.', '').replace(',', '.')
            return float(cleaned)
        except (ValueError, AttributeError):
            return 0.0

    def _calculate_score(self, contract: Dict[str, Any]) -> int:
        """
        Calcular score de relevância baseado em keywords
        """
        designation = contract.get('contractDesignation', '').lower()
        entity = contract.get('contractingEntity', '').lower()

        score = 0

        # +10 pontos por cada keyword que apareça
        for keyword in self.keywords:
            if keyword in designation or keyword in entity:
                score += 10

        # Opcional: boost para contratos de maior valor
        # price = self._parse_price(contract.get('basePrice', '0'))
        # if price > 50000:
        #     score += 5

        return score

    def _should_exclude(self, contract: Dict[str, Any]) -> bool:
        """
        Verificar se o contrato contém termos de exclusão
        """
        designation = contract.get('contractDesignation', '').lower()

        for term in self.exclude_terms:
            if term in designation:
                return True

        return False

    def _matches_price_range(self, contract: Dict[str, Any]) -> bool:
        """
        Verificar se o preço está dentro do intervalo definido
        """
        price = self._parse_price(contract.get('basePrice', '0'))
        return self.min_price <= price <= self.max_price

    def filter_and_classify(self, contracts: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        Filtrar e classificar lista de contratos

        Returns:
            Lista de contratos relevantes, ordenados por score (maior primeiro)
        """
        results = []

        for contract in contracts:
            # Aplicar filtros de exclusão
            if self._should_exclude(contract):
                continue

            # Verificar preço
            if not self._matches_price_range(contract):
                continue

            # Calcular relevância
            score = self._calculate_score(contract)

            # Apenas manter se tem pelo menos alguma relevância
            if score > 0:
                contract['relevanceScore'] = score
                results.append(contract)

        # Ordenar por score (maior primeiro)
        results.sort(key=lambda x: x['relevanceScore'], reverse=True)

        logger.info(f"Filtrados {len(results)} contratos relevantes de {len(contracts)} totais")

        return results

File: src/services/contractFilter.ts
import { Contract } from './baseGovService';

export interface UserProfile {
    keywords: string[];
    minPrice?: number;
    maxPrice?: number;
    excludedTerms?: string[];
}

export interface ClassifiedContract extends Contract {
    relevanceScore: number;
    matchedKeywords: string[];
}

export class ContractFilter {
    private profile: UserProfile;

    constructor(profile: UserProfile) {
        this.profile = profile;
    }

    public filterAndClassify(contracts: Contract[]): ClassifiedContract[] {
        return contracts
            .filter(c => this.passesHardFilters(c))
            .map(c => this.classify(c))
            .filter(c => c.relevanceScore > 0) // Isto pode ser chato - mantém apenas os que têm pelo menos um ponto
            .sort((a, b) => b.relevanceScore - a.relevanceScore);
    }

    private passesHardFilters(contract: Contract): boolean {
        // Separar o preço "75.000,00 €" -> 75000.00
        const priceStr = contract.basePrice.replace(/\./g, '').replace(',', '.').replace(/[^0-9.]/g, '');
        const price = parseFloat(priceStr);

        if (this.profile.minPrice && price < this.profile.minPrice) return false;
        if (this.profile.maxPrice && price > this.profile.maxPrice) return false;

        if (this.profile.excludedTerms && this.profile.excludedTerms.some(term =>
            contract.contractDesignation.toLowerCase().includes(term.toLowerCase()))) {
            return false;
        }

        return true;
    }

    private classify(contract: Contract): ClassifiedContract {
        const text = contract.contractDesignation.toLowerCase();
        const matches: string[] = [];
        let score = 0;

        for (const keyword of this.profile.keywords) {
            if (text.includes(keyword.toLowerCase())) {
                matches.push(keyword);
                score += 10; // Pontos base para correspondência de keywords
            }
        }

        // Mais guita vale mais pontos (log opcional)
        // const price = parseFloat(contract.basePrice.replace(/\./g, '').replace(',', '.').replace(/[^0-9.]/g, ''));
        // if (price > 50000) score += 5;

        return {
            ...contract,
            relevanceScore: score,
            matchedKeywords: matches
        };
    }
}
